dartagnan/src/main/java/com/dat3m/dartagnan/analysis/Refinement.java

```diff
@@ -0,0 +1,387 @@
+                // We couldn't verify the found counterexample, nor exclude it.
+                System.out.println("PROCEDURE was inconclusive.");
+                return UNKNOWN;
+            } else if (isSat) {
+                // We found a true violation
+                System.out.println("Violation verified.");
+            } else {
+                // We showed safety but still need to verify bounds
+                System.out.println("Bounded safety proven.");
+            }
+        } else {
+            if (isSat && res == UNKNOWN) {
+                return UNKNOWN;
+            }
+        }
+
+        long boundCheckTime = 0;
+        if (prover.isUnsat()) {
+            // ------- CHECK BOUNDS -------
+            lastTime = System.currentTimeMillis();
+            prover.pop();
+            prover.addConstraint(bmgr.not(program.encodeNoBoundEventExec(ctx)));
+            res = !prover.isUnsat() ? UNKNOWN : PASS; // Initial bound check without any WMM constraints
+            if (res == UNKNOWN) {
+                //TODO: This is just a temporary fallback
+                // We probably have to perform a second refinement for the bound checks!
+                for (DNF<CoreLiteral> violation : foundViolations) {
+                    refine(prover, ctx, violation, perms);
+                }
+                res = !prover.isUnsat() ? UNKNOWN : PASS;
+            }
+            boundCheckTime = System.currentTimeMillis() - lastTime;
+        } else {
+            res = FAIL;
+        }
+        if (REF_PRINT_STATISTICS) {
+            printSummary(statList, totalSolvingTime, boundCheckTime);
+        }
+
+        res = program.getAss().getInvert() ? res.invert() : res;
+        logger.info("Verification finished with result " + res);
+        return res;
+    }
+
+
+    // This method adds new constraints to the prover based on the found violations.
+    // Furthermore, it computes symmetric violations if symmetry learning is enabled.
+    private static void refine(ProverEnvironment prover, SolverContext ctx, DNF<CoreLiteral> coreViolations,
+                               List<Function<Event, Event>> perms) throws InterruptedException {
+        BooleanFormulaManager bmgr = ctx.getFormulaManager().getBooleanFormulaManager();
+        for (Function<Event, Event> p : perms) {
+            BooleanFormula refinement = bmgr.makeTrue();
+            for (Conjunction<CoreLiteral> violation : coreViolations.getCubes()) {
+                BooleanFormula clause = bmgr.makeFalse();
+                for (CoreLiteral literal : violation.getLiterals()) {
+                    clause = bmgr.or(clause, bmgr.not(permuteAndConvert(literal, p, ctx)));
+                }
+                refinement = bmgr.and(refinement, clause);
+            }
+            prover.addConstraint(refinement);
+        }
+    }
+
+
+    // ======================= Helper Methods ======================
+
+    // -------------------- Printing -----------------------------
```

Comment:
I added a custom logger for all Refinement output

Reply:
Not sure the current output format os the best is we want to use these logs to automatically extract information to create tables/graphs, but this is something we can improve later. It is good that you created a dedicated log file for refinement (have not checked how having two consoles configurations work)

Reply:
The logging is exclusively into the new refinement log file. Everything that gets logged there will not get logged in the regular log files. Getting a nice logging output format is difficult for multi-line outputs. Originally, I replaced every println by a call to the logger, but that is not only ugly, it is also conceptually wrong. So I replaced all this stuff with the string builder and a separate formatter. Now between every two log outputs something will actually have happenend in the code.

Reply:
If we put the code `logger.info()` in the correct part of the code (which might not match the position where the string builders are placed right now), we should obtain the same order as we currently have, don't we?

Reply:
What do you mean by correct part of the code and order? It is not about the order of logging but about the number of logging events. If I don't use string builders for multiline messages, I get the logging prefix for each and every line of output. This is conceptually wrong, since I want to log a single event which just happens to have a lot of information.
Btw. I will push a reworked version now, which has more fine grained control over the logging levels (TRACE prints everything, DEBUG prints the statistics per iteration and INFO only prints the summary statistics as well as some other minor stuff).

Reply:
I don't understand why you think this is **conceptually wrong**. Why e.g.
```
.append("Total solving time( ms): ").append(totalSolvingTime).append("\n")
.append("Total model construction time( ms): ").append(totalModelTime).append("\n");
```
have to be part of the same event?

For me actually, it is exactly the contrary. They are conceptually two different things (one is about solving time, other is about model construction time), thus they should be different events.

I agree that this would result in having something like
```
10:03:07 [INFO ] Refinement.logSummary - Total solving time( ms): X
10:03:07 [INFO ] Refinement.logSummary - Total model construction time( ms): Y

```
in the log, but I don't see why this is a problem

Reply:
Well, the content of each line may be related to something different and in this case you might argue that you can log everything separately. But I also don't see the benefit of having the same logging prefix over and over again given the fact that everything is logged at once there. On the other hand, I also log the ```RefinementStats``` object which is a single object and hence warrants a single log entry. The only difference is that I don't have a single object representing the summary statistics but I still would want to log both the same way. In terms of reading and analysing the logs, I think it should be even easier if I can just look for the log entry corresponding to "Summary". Especially when debugging, it is much harder to parse the console output if not every single line has this logging prefix (mostly with identical timestamps!). 
For example, right now if I enable ```INFO``` level debugging, I get exactly 4 logging entries:
- Start of refinement
- End of refinement (with intermediate result)
- Summary
- Final verification result

I like this quite more than having something like 10 entries just for the summary. I would agree with you, if the different parts of the procedure (solving, model construction etc.) were logging themselves, but right now only the entry point ```Refinement``` performs logging and mostly about statistics rather than functional steps.

Reply:
I agree that depending on the use of the log, each format has it benefit. You have been used this quite a lot for debugging (which I never did) and I believe that for that purpose, the current output makes it easier.

I was thinking more about using the log to automatically extract information (using scripts) to generate tables and graphics and there I think having a line per item is the way to go.

Maybe we can use different levels to capture different purposes (unfortunately there is no `STATS` level or something of the sort).

Reply:
In what way do separately logged lines help you in processing, if each of their prefixes is identical? The prefixes would not help you at all in distuingishing anything. Wouldn't it be easier to search for the log entry (which is called "Summary" right now) and then process the following lines?
Regarding the logging levels, you can define custom logging levels like ```STATS``` so that is no problem.
Either way, I would revisit the logging format once we actually want to automatically process log data. Right now, we can only speculate what the easier way would be.

Reply:
Yes, agree this is mostly speculation and we can revisit this when we need it.